### PR TITLE
Niad 313 soap service 2nd version

### PR DESCRIPTION
**Pull request by Alicja-Kainos**

### CODE REVIEW

File: service/build.gradle [hunk 1]
```diff
@@ -22,7 +22,8 @@ dependencies {
     implementation "ca.uhn.hapi.fhir:hapi-fhir-structures-dstu3:4.2.0"
     implementation "uk.nhs.connect.iucds:iucds-schema:3.0.RC1.2"
     implementation "org.springframework.boot:spring-boot-starter-actuator:2.3.1.RELEASE"
-
+    implementation group: 'dom4j', name: 'dom4j', version: '1.6.1'
```
Comment (dawidskainos) [inline service/build.gradle:25]: Can we please stick to the convention and use colons?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java [hunk 2]
```diff
@@ -29,13 +38,32 @@
 
     @PostMapping(value = "/report", consumes = { TEXT_XML_VALUE, APPLICATION_XML_VALUE })
     @ResponseStatus(value = ACCEPTED)
-    public void postReport(@RequestBody String reportXml) throws JsonProcessingException {
+    public void postReport(@RequestBody String reportXml) {
         try {
-            POCDMT000002UK01ClinicalDocument1 clinicalDocument = extractClinicalDocument(reportXml);
+            SAXReader reader = new SAXReader();
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:43]: Can we extract the parsing part to a separate class? Keep the controller as clean and simple as possible?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java [hunk 3]
```diff
@@ -29,13 +38,32 @@
...
             encounterReportService.transformAndPopulateToGP(clinicalDocument);
-        } catch (XmlException e) {
+        } catch (XmlException | DocumentException e) {
             throw new ResponseStatusException(BAD_REQUEST, e.getMessage());
         }
     }
+
+    private String getMessageId(Document document) {
+        Node headerNode = document.selectSingleNode("//*[local-name()='Header']");
+        Node messageIdNode = headerNode.selectSingleNode("//*[local-name()='MessageID']");
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:61]: Isn't it enough to search for MessageID element in RootElement?

Comment (Alicja-Kainos) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:61]: it is :)

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java (hunk 2 above)
Comment (Alicja-Kainos) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:43]: done

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportService.java [hunk 4]
```diff
@@ -21,10 +23,14 @@
...
     private AmqpProperties amqpProperties;
 
-    public void transformAndPopulateToGP(POCDMT000002UK01ClinicalDocument1 clinicalDocumentDocument) {
+    public void transformAndPopulateToGP(POCDMT000002UK01ClinicalDocument1 clinicalDocumentDocument, String messageId) {
         Bundle encounterBundle = encounterReportBundleService.createEncounterBundle(clinicalDocumentDocument);
 
-        jmsTemplate.send(amqpProperties.getQueueName(), session -> session.createTextMessage(toJsonString(encounterBundle)));
+        jmsTemplate.send(amqpProperties.getQueueName(), session -> {
+            TextMessage message = session.createTextMessage(toJsonString(encounterBundle));
+            message.setStringProperty("messageId", messageId);
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportService.java:31]: Can we extract "messageId" to a constant?

File: service/src/test/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportControllerTest.java [hunk 5]
```diff
@@ -32,20 +33,20 @@
...
     @Test
-    public void postReportValidRequest() throws JsonProcessingException {
+    public void postReportValidRequest() {
         String validRequest = getValidReportRequest();
 
         reportController.postReport(validRequest);
 
         ArgumentCaptor<POCDMT000002UK01ClinicalDocument1> captor = ArgumentCaptor.forClass(POCDMT000002UK01ClinicalDocument1.class);
-        verify(encounterReportService).transformAndPopulateToGP(captor.capture());
+        verify(encounterReportService).transformAndPopulateToGP(captor.capture(), ArgumentMatchers.anyString());
```
Comment (dawidskainos) [inline service/src/test/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportControllerTest.java:42]: Can we use the exact messageId, not anyString?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        SAXReader reader = new SAXReader();
+        Document document = reader.read(new StringReader(reportXml));
+
+        reportElementsMap.put(ReportElement.MESSAGE_ID, getMessageId(document));
+        reportElementsMap.put(ReportElement.DISTRIBUTION_ENVELOPE, getDistributionEnvelope(document));
+        return reportElementsMap;
+    }
+
+    private static String getMessageId(Document document) {
+        Node messageIdNode = document.selectSingleNode("//*[local-name()='MessageID']");
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java:32]: can we extract xpaths to constants?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Map;
+
+import org.dom4j.Document;
+import org.dom4j.DocumentException;
+import org.dom4j.Node;
+import org.dom4j.io.SAXReader;
+
+public class ReportParserUtil {
+
+    public enum ReportElement{
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java:14]: I wouldn't make it an inner class. Can we extract it? Also, please remember about formatting. There's a missing space after the class name.

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.Map;
+
+import org.dom4j.Document;
+import org.dom4j.DocumentException;
+import org.dom4j.Node;
+import org.dom4j.io.SAXReader;
+
+public class ReportParserUtil {
+
+    public static final String MESSAGE_ID_NODE = "//*[local-name()='MessageID']";
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java:14]: should be private

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportService.java [hunk 9]
```diff
@@ -13,6 +15,8 @@
 @AllArgsConstructor
 public class EncounterReportService {
 
+    public static final String MESSAGE_ID = "messageId";
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportService.java:18]: should be private

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java (hunk 8 above)
Comment (Alicja-Kainos) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java:14]: ok

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportService.java (hunk 9 above)
Comment (Alicja-Kainos) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/service/EncounterReportService.java:18]: ok

**Pull request merged**